cpp/tests/neighbors/ann_cagra_ace.cuh

```diff
@@ -0,0 +1,275 @@
+                                    ps.n_queries,
+                                    ps.k,
+                                    0.003,
+                                    ps.min_recall))
+          << "Disk-based ACE index loaded via HNSW failed recall check";
+      } else {
+        // For in-memory ACE, we can search directly
+        EXPECT_FALSE(index.on_disk());
+        ASSERT_GT(index.graph().size(), 0);
+        EXPECT_EQ(index.graph_degree(), 64);
+
+        rmm::device_uvector<DistanceT> distances_dev(queries_size, stream_);
+        rmm::device_uvector<IdxT> indices_dev(queries_size, stream_);
+
+        auto queries_view = raft::make_device_matrix_view<const DataT, int64_t>(
+          search_queries.data(), ps.n_queries, ps.dim);
+        auto indices_view =
+          raft::make_device_matrix_view<IdxT, int64_t>(indices_dev.data(), ps.n_queries, ps.k);
+        auto distances_view = raft::make_device_matrix_view<DistanceT, int64_t>(
+          distances_dev.data(), ps.n_queries, ps.k);
+
+        cagra::search_params search_params;
+        search_params.itopk_size = 64;
+
+        cagra::search(handle_, search_params, index, queries_view, indices_view, distances_view);
+
+        raft::update_host(distances_ace.data(), distances_dev.data(), queries_size, stream_);
+        raft::update_host(indices_ace.data(), indices_dev.data(), queries_size, stream_);
+        raft::resource::sync_stream(handle_);
+
+        EXPECT_TRUE(eval_neighbours(indices_naive,
+                                    indices_ace,
+                                    distances_naive,
+                                    distances_ace,
+                                    ps.n_queries,
+                                    ps.k,
+                                    0.003,
+                                    ps.min_recall))
+          << "In-memory ACE index failed recall check";
+      }
+    }
+
+    // Clean up temporary directory
+    std::filesystem::remove_all(temp_dir);
+  }
+
+  void SetUp() override
+  {
+    database_dev.resize(((size_t)ps.n_rows) * ps.dim, stream_);
+    search_queries.resize(ps.n_queries * ps.dim, stream_);
+    raft::random::RngState r(1234ULL);
+    InitDataset(handle_, database_dev.data(), ps.n_rows, ps.dim, ps.metric, r);
+    InitDataset(handle_, search_queries.data(), ps.n_queries, ps.dim, ps.metric, r);
+    raft::resource::sync_stream(handle_);
+  }
+
+  void TearDown() override
+  {
+    raft::resource::sync_stream(handle_);
+    database_dev.resize(0, stream_);
+    search_queries.resize(0, stream_);
+  }
+
+ private:
+  raft::resources handle_;
+  rmm::cuda_stream_view stream_;
+  AnnCagraAceInputs ps;
+  rmm::device_uvector<DataT> database_dev;
+  rmm::device_uvector<DataT> search_queries;
+};
+
+inline std::vector<AnnCagraAceInputs> generate_ace_inputs()
+{
+  return raft::util::itertools::product<AnnCagraAceInputs>(
+    {10},           // n_queries
+    {5000},         // n_rows
```

Comment:
What is the smallest number of rows `bulid_ace` can handle? When cuvs is integrated into other libs, their unit test might call it with very low values (e.g. n_rows=10). Will `build_ace` work in that case or throw a useful error message?

Reply:
The C test uses four elements which should be fine once the `from_cagra` issue with odd graph degrees is fixed. `build_ace` expects (and forces) at least two partitions for meaningful results. I will test this once we have the `from_cagra` fix and throw a error message in case of too few elements.